### PR TITLE
DialogManager

### PR DESCRIPTION
**Pull request by axelsrz**
DialogManager support

**This PR is ready for review but awaiting for more unit tests**

### CODE REVIEW

**Review by EricDahlvang — COMMENTED** [review]
I did not get through the entire PR.  I think this is a little outdated from the current c# DialogManager.  Please let me know when you're ready, and i'll make another pass.

File: libraries/botbuilder-dialogs/botbuilder/dialogs/dialog.py [hunk 1]
```diff
@@ -105,3 +106,83 @@ async def end_dialog(  # pylint: disable=unused-argument
...
+        self, path: str, line_number: int
+    ):  # pylint: disable=unused-argument
+        """
+        Registers a SourceRange in the provided location.
+        :param path: The path to the source file.
+        :param line_number: The line number where the source will be located on the file.
+        :return:
+        """
+        if path:
+            # This will be added when debbuging support is ported.
```
[the remark this answers is not in the document]
Comment (axelsrz) [reply libraries/botbuilder-dialogs/botbuilder/dialogs/dialog.py:182]: Just opened [one](https://github.com/microsoft/botbuilder-python/issues/1472) for tracking.

File: libraries/botbuilder-dialogs/botbuilder/dialogs/dialog_manager.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+        self.expire_after: int = None
+
+    async def on_turn(self, context: TurnContext) -> DialogManagerResult:
+        """
+        Runs dialog system in the context of an ITurnContext.
+        :param context: turn context.
+        :return:
+        """
+        # pylint: disable=too-many-statements
+        # Lazy initialize RootDialog so it can refer to assets like LG function templates
```
[the remark this answers is not in the document]
Comment (axelsrz) [reply libraries/botbuilder-dialogs/botbuilder/dialogs/dialog_manager.py:81]: This line was relevant at the time of this [PR](https://github.com/microsoft/botbuilder-python/pull/1409), this [issue ](https://github.com/microsoft/botbuilder-python/issues/1473) will track the drift.

File: libraries/botbuilder-dialogs/botbuilder/dialogs/dialog_context.py [hunk 3]
```diff
@@ -57,12 +65,33 @@ def active_dialog(self):
...
+        :return DialogContext:
+        """
+        # pylint: disable=import-outside-toplevel
+        instance = self.active_dialog
+
+        if instance:
+            dialog = self.find_dialog_sync(instance.id)
+
+            # This import prevents circular dependency issues
+            from .dialog_container import DialogContainer
```
[the remark this answers is not in the document]
Comment (axelsrz) [reply libraries/botbuilder-dialogs/botbuilder/dialogs/dialog_context.py:88]: There is a circular dependency by design. In the DialogContainer class, methods such as create_child_context, on_dialog_event and check_for_version_change_async (based on their C# counterparts) receive a DialogContext object as parameter.

**Review by tracyboehrer — APPROVED** [review]

**Pull request merged**